test-loop-tests/src/tests/global_contracts_distribution.rs

```diff
@@ -0,0 +1,186 @@
+use std::collections::BTreeMap;
+use std::sync::Arc;
+
+use itertools::Itertools;
+use near_async::time::Duration;
+use near_chain_configs::test_genesis::{TestEpochConfigBuilder, ValidatorsSpec};
+use near_client::Client;
+use near_o11y::testonly::init_test_logger;
+use near_primitives::action::{GlobalContractDeployMode, GlobalContractIdentifier};
+use near_primitives::block::MaybeNew;
+use near_primitives::epoch_manager::EpochConfigStore;
+use near_primitives::receipt::ReceiptEnum;
+use near_primitives::shard_layout::ShardLayout;
+use near_primitives::types::{AccountId, BlockHeightDelta};
+use near_primitives::version::PROTOCOL_VERSION;
+use near_vm_runner::ContractCode;
+
+use crate::setup::builder::TestLoopBuilder;
+use crate::setup::env::TestLoopEnv;
+use crate::utils::setups::derive_new_epoch_config_from_boundary;
+use crate::utils::transactions::{check_txs, deploy_global_contract, use_global_contract};
+use crate::utils::{ONE_NEAR, get_node_client, run_until_node_head_height};
+
+const EPOCH_LENGTH: BlockHeightDelta = 5;
+
+// TODO(stedfn): remove "nightly" feature once we have a new protocol version
+// (see the explanation for `slow_test_resharding_v3_global_contract_by_hash`).
+#[test]
+#[cfg_attr(not(feature = "nightly"), ignore)]
+fn test_global_receipt_distribution_at_resharding_boundary() {
+    let mut env = GlobalContractsReshardingTestEnv::setup();
+    let expected_new_shard_layout_height = EPOCH_LENGTH * 2 + 2;
+    // This height is picked so that the first global contract distribution receipt reaches
+    // shard that is being split at the first height after the resharding
+    let send_deploy_tx_height = expected_new_shard_layout_height - 3;
+
+    run_until_node_head_height(
+        &mut env.env,
+        &env.chunk_producer,
+        send_deploy_tx_height,
+        Duration::seconds(10),
+    );
+    assert_eq!(env.current_shard_layout(), env.base_shard_layout);
+
+    // Deploying global contract with the user from the split shard.
+    // The first target shard for the global contract distribution receipt
+    // is user's shard, this way we ensure that we hit the shard that is
+    // split at resharding.
+    let deploy_user = env.users[0].clone();
```

Comment:
mini nit: users[0] being in the parent shard is a bit implicit, maybe make it into proper structure? Very minor, feel free to ignore.